In [20]:
import numpy as np
import pandas as pd
import yaml
import os
from tqdm import tqdm
import pickle

In [2]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [3]:
filenames = []
for file in os.listdir('D:\SEM4\Project\ipl'):
    filenames.append(os.path.join('D:\SEM4\Project\ipl',file))

Uploaingd the yaml file set.

In [4]:
filenames[0:5]

['D:\\SEM4\\Project\\ipl\\1082591.yaml',
 'D:\\SEM4\\Project\\ipl\\1082592.yaml',
 'D:\\SEM4\\Project\\ipl\\1082593.yaml',
 'D:\\SEM4\\Project\\ipl\\1082594.yaml',
 'D:\\SEM4\\Project\\ipl\\1082595.yaml']

uploading yaml data file into dataframe and intriduse the new match id columns.

In [5]:
final_df = pd.DataFrame()
for i, file in enumerate(tqdm(filenames)):
    with open(file, 'r') as f:
        data = yaml.safe_load(f)
        df = pd.json_normalize(data)
        df['match_id'] = i+1
        #final_df = final_df.append(df)
        final_df = pd.concat([final_df, df])

final_df

100%|█████████▉| 950/951 [03:07<00:00,  5.07it/s]


NotImplementedError: 

In [7]:
Final_backup = final_df.copy()


In [8]:
final_df

innings  meta.data_version  \
0   [{'1st innings': {'team': 'Sunrisers Hyderabad...               0.91   
0   [{'1st innings': {'team': 'Mumbai Indians', 'd...               0.91   
0   [{'1st innings': {'team': 'Gujarat Lions', 'de...               0.91   
0   [{'1st innings': {'team': 'Rising Pune Supergi...               0.91   
0   [{'1st innings': {'team': 'Royal Challengers B...               0.91   
..                                                ...                ...   
0   [{'1st innings': {'team': 'Delhi Daredevils', ...               0.91   
0   [{'1st innings': {'team': 'Gujarat Lions', 'de...               0.91   
0   [{'1st innings': {'team': 'Sunrisers Hyderabad...               0.91   
0   [{'1st innings': {'team': 'Gujarat Lions', 'de...               0.91   
0   [{'1st innings': {'team': 'Sunrisers Hyderabad...               0.91   

   meta.created  meta.revision  info.balls_per_over  info.city  \
0    2017-04-06              1                    6  Hyderabad   
0    2017-04-07              1                    6       Pune   
0    2017-04-07              2                    6     Rajkot   
0    2017-04-08              1                    6     Indore   
0    2017-04-08              2                    6  Bengaluru   
..          ...            ...                  ...        ...   
0    2016-05-23              1                    6     Raipur   
0    2016-05-24              1                    6  Bangalore   
0    2016-05-25              1                    6      Delhi   
0    2016-05-28              1                    6      Delhi   
0    2016-05-30              1                    6  Bangalore   

   info.competition    info.dates info.gender info.match_type  ...  \
0               IPL  [2017-04-05]        male             T20  ...   
0               IPL  [2017-04-06]        male             T20  ...   
0               IPL  [2017-04-07]        male             T20  ...   
0               IPL  [2017-04-08]        male             T20  ...   
0               IPL  [2017-04-08]        male             T20  ...   
..              ...           ...         ...             ...  ...   
0               IPL  [2016-05-22]        male             T20  ...   
0               IPL  [2016-05-24]        male             T20  ...   
0               IPL  [2016-05-25]        male             T20  ...   
0               IPL  [2016-05-27]        male             T20  ...   
0               IPL  [2016-05-29]        male             T20  ...   

    info.registry.people.GS Sandhu info.players.Rising Pune Supergiants  \
0                              NaN                                  NaN   
0                              NaN                                  NaN   
0                              NaN                                  NaN   
0                              NaN                                  NaN   
0                              NaN                                  NaN   
..                             ...                                  ...   
0                              NaN                                  NaN   
0                              NaN                                  NaN   
0                              NaN                                  NaN   
0                              NaN                                  NaN   
0                              NaN                                  NaN   

    info.registry.people.P Sahu info.registry.people.KJ Abbott  \
0                           NaN                            NaN   
0                           NaN                            NaN   
0                           NaN                            NaN   
0                           NaN                            NaN   
0                           NaN                            NaN   
..                          ...                            ...   
0                           NaN                            NaN   
0                           NaN                            NaN   
0   

checking all columns

In [9]:
column_list = list(final_df.keys())
print(column_list)

['innings', 'meta.data_version', 'meta.created', 'meta.revision', 'info.balls_per_over', 'info.city', 'info.competition', 'info.dates', 'info.gender', 'info.match_type', 'info.outcome.by.runs', 'info.outcome.winner', 'info.overs', 'info.player_of_match', 'info.players.Royal Challengers Bangalore', 'info.players.Sunrisers Hyderabad', 'info.registry.people.A Choudhary', 'info.registry.people.A Deshmukh', 'info.registry.people.A Nehra', 'info.registry.people.AY Dandekar', 'info.registry.people.B Kumar', 'info.registry.people.BCJ Cutting', 'info.registry.people.Bipul Sharma', 'info.registry.people.CH Gayle', 'info.registry.people.CJ Jordan', 'info.registry.people.DA Warner', 'info.registry.people.DJ Hooda', 'info.registry.people.J Srinath', 'info.registry.people.KM Jadhav', 'info.registry.people.MC Henriques', 'info.registry.people.Mandeep Singh', 'info.registry.people.N Pandit', 'info.registry.people.NJ Llong', 'info.registry.people.NV Ojha', 'info.registry.people.Rashid Khan', 'info.regi

selecting the required columns and droping other columns

In [10]:

columns_to_keep = ['innings', 'info.dates','info.gender', 'info.match_type','info.competition','info.outcome.winner','info.outcome.by.runs','info.overs','info.player_of_match', 'info.teams','info.toss.decision', 'info.toss.winner','info.umpires', 'info.venue','info.city', 'match_id']
selected_columns = final_df[columns_to_keep]


final_df_new = final_df.drop([col for col in final_df.columns if col not in columns_to_keep], axis=1)

chekcing new dataframe which has total 950 row and 16 columns. this dataset shows all the ipl matches data which are total 950 matches are played between year of 2008 to 2022. the ball by ball daya is in json direcoty formate under the innings colums. 

In [11]:
final_df_new

innings  info.city  \
0   [{'1st innings': {'team': 'Sunrisers Hyderabad...  Hyderabad   
0   [{'1st innings': {'team': 'Mumbai Indians', 'd...       Pune   
0   [{'1st innings': {'team': 'Gujarat Lions', 'de...     Rajkot   
0   [{'1st innings': {'team': 'Rising Pune Supergi...     Indore   
0   [{'1st innings': {'team': 'Royal Challengers B...  Bengaluru   
..                                                ...        ...   
0   [{'1st innings': {'team': 'Delhi Daredevils', ...     Raipur   
0   [{'1st innings': {'team': 'Gujarat Lions', 'de...  Bangalore   
0   [{'1st innings': {'team': 'Sunrisers Hyderabad...      Delhi   
0   [{'1st innings': {'team': 'Gujarat Lions', 'de...      Delhi   
0   [{'1st innings': {'team': 'Sunrisers Hyderabad...  Bangalore   

   info.competition    info.dates info.gender info.match_type  \
0               IPL  [2017-04-05]        male             T20   
0               IPL  [2017-04-06]        male             T20   
0               IPL  [2017-04-07]        male             T20   
0               IPL  [2017-04-08]        male             T20   
0               IPL  [2017-04-08]        male             T20   
..              ...           ...         ...             ...   
0               IPL  [2016-05-22]        male             T20   
0               IPL  [2016-05-24]        male             T20   
0               IPL  [2016-05-25]        male             T20   
0               IPL  [2016-05-27]        male             T20   
0               IPL  [2016-05-29]        male             T20   

    info.outcome.by.runs          info.outcome.winner  info.overs  \
0                   35.0          Sunrisers Hyderabad          20   
0                    NaN       Rising Pune Supergiant          20   
0                    NaN        Kolkata Knight Riders          20   
0                    NaN              Kings XI Punjab          20   
0                   15.0  Royal Challengers Bangalore          20   
..                   ...                          ...         ...   
0                    NaN  Royal Challengers Bangalore          20   
0                    NaN  Royal Challengers Bangalore          20   
0                   22.0          Sunrisers Hyderabad          20   
0                    NaN          Sunrisers Hyderabad          20   
0                    8.0          Sunrisers Hyderabad          20   

   info.player_of_match                                         info.teams  \
0        [Yuvraj Singh]  [Sunrisers Hyderabad, Royal Challengers Bangal...   
0           [SPD Smith]           [Rising Pune Supergiant, Mumbai Indians]   
0             [CA Lynn]             [Gujarat Lions, Kolkata Knight Riders]   
0          [GJ Maxwell]          [Kings XI Punjab, Rising Pune Supergiant]   
0           [KM Jadhav]    [Royal Challengers Bangalore, Delhi Daredevils]   
..                  ...                                                ...   
0             [V Kohli]    [Delhi Daredevils, Royal Challengers Bangalore]   
0      [AB de Villiers]       [Gujarat Lions, Royal Challengers Bangalore]   
0        [MC Henriques]       [Sunrisers Hyderabad, Kolkata Knight Riders]   
0           [DA Warner]               [Gujarat Lions, Sunrisers Hyderabad]   
0         [BCJ Cutting]  [Royal Challengers Bangalore, Sunrisers Hydera...   

   info.toss.decision             info.toss.winner  \
0               field  Royal Challengers Bangalore   
0               field       Rising Pune Supergiant   
0               field        Kolkata Knight Riders   
0               field              Kings XI Punjab   
0                 bat  Royal Challengers Bangalore   
..                ...                          ...   
0               field  Royal Challengers Bangalore   
0               field  Royal Challengers Bangalore   
0               field        Kolkata Knight Riders   
0               field          Sunrisers Hyderabad   
0                 bat          Sunrisers Hyderabad   

                       info.umpir

In [12]:
final_df_new['info.overs'].value_counts()

20    950
Name: info.overs, dtype: int64

creating dump backup file.

In [ ]:

pickle.dump(final_df_new,open('dataset_level1.pkl','wb'))

save final matches dataset into  new dataset call match_data.

In [13]:
match_data = final_df_new

exploring the innings columns data which we are going to create new dataset for our project.

the innings column shows the data each match all innings ball by ball data that cover who is th bowler, who is the batsmen, the run  etc.

In [14]:

match_data.iloc[0]['innings']

[{'1st innings': {'team': 'Sunrisers Hyderabad',
   'deliveries': [{0.1: {'batsman': 'DA Warner',
      'bowler': 'TS Mills',
      'non_striker': 'S Dhawan',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.2: {'batsman': 'DA Warner',
      'bowler': 'TS Mills',
      'non_striker': 'S Dhawan',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.3: {'batsman': 'DA Warner',
      'bowler': 'TS Mills',
      'non_striker': 'S Dhawan',
      'runs': {'batsman': 4, 'extras': 0, 'total': 4}}},
    {0.4: {'batsman': 'DA Warner',
      'bowler': 'TS Mills',
      'non_striker': 'S Dhawan',
      'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
    {0.5: {'batsman': 'DA Warner',
      'bowler': 'TS Mills',
      'extras': {'wides': 2},
      'non_striker': 'S Dhawan',
      'runs': {'batsman': 0, 'extras': 2, 'total': 2}}},
    {0.6: {'batsman': 'S Dhawan',
      'bowler': 'TS Mills',
      'non_striker': 'DA Warner',
      'runs': {'batsman': 0, 'extras': 0, 'tot

creating new dataframe called delivery data in this data we are going to create new columns deliveries, batsman, totalrun etc. also we collecting the match id and date information from the $match$ dataset.

This code creates an empty list called delivery_data_list and then iterates over each row in the match_data DataFrame. For each row, it extracts several pieces of information from the row and its nested dictionaries and lists, and creates a new dictionary with this information. It then appends this dictionary to a list called temp_data. After the innermost loop finishes, it converts temp_data to a DataFrame, appends it to delivery_data_list, and then continues the loop. Finally, it concatenates all the DataFrames in delivery_data_list using pd.concat() and assigns the result to the delivery_data DataFrame.

In [15]:
delivery_data_list = []
for index, row in match_data.iterrows():
    match_id = row['match_id']
    info_dates = row['info.dates'][0]
    for object in row['innings']:
        for inningKey in object: 
            deliveries = object[inningKey]['deliveries']
            teams = row['info.teams']
            batting_team = object[inningKey]['team']
            city = row['info.city']
            venue = row['info.venue']
            temp_data = []
            for delivery in deliveries:
                for key in delivery.keys():
                    ball_of_match = key
                    batsman = delivery[key]['batsman']
                    non_striker = delivery[key]['non_striker']
                    bowler = delivery[key]['bowler']
                    total_runs = delivery[key]['runs']['total']
                    bat_runs = delivery[key]['runs']['batsman']
                    extra_runs = delivery[key]['runs']['extras']
                    player_of_dismissed = delivery[key].get('wicket', {}).get('player_out', '0')
                    wicket_type = delivery[key].get('wicket', {}).get('kind', '0')
                    fielder = delivery[key].get('wicket', {}).get('fielders', '0')
                    extra_run_type = list(delivery[key].get('extras', {}).keys())
                    extra_run_type = extra_run_type and extra_run_type[0] or extra_run_type

                    temp_data.append({
                        'match_id': match_id,
                        'innings':inningKey,
                        'info_Dates': info_dates,
                        'teams': teams,
                        'batting_team': batting_team,
                        'ball': ball_of_match,
                        'batsman': batsman,
                        'non_striker':non_striker,
                        'bowler': bowler,
                        'total_runs': total_runs,
                        'bat_runs': bat_runs,
                        'extra_runs': extra_runs,
                        'extra_run_type':extra_run_type,
                        'player_dismissed': player_of_dismissed,
                        'wicket_type':wicket_type,
                        'fielder': fielder,
                        'city': city,
                        'venue': venue
                    })
            delivery_data_list.append(pd.DataFrame(temp_data))
delivery_data = pd.concat(delivery_data_list, ignore_index=True)


In [ ]:
# delivery_data = pd.DataFrame()
# for index, row in match_data.iterrows():
#     match_id = row['match_id']
#     info_dates = row['info.dates'][0]
#     for object in row['innings']:
#         for inningKey in object: 
#             deliveries = object[inningKey]['deliveries']
#             teams = row['info.teams']
#             batting_team = object[inningKey]['team']
#             city = row['info.city']
#             venue = row['info.venue']

#             for delivery in deliveries:
#                 for key in delivery.keys():
#                     ball_of_match = key
#                     batsman = delivery[key]['batsman']
#                     non_striker = delivery[key]['non_striker']
#                     bowler = delivery[key]['bowler']
#                     total_runs = delivery[key]['runs']['total']
#                     bat_runs = delivery[key]['runs']['batsman']
#                     extra_runs = delivery[key]['runs']['extras']
#                     player_of_dismissed = delivery[key].get('wicket', {}).get('player_out', '0')
#                     wicket_type = delivery[key].get('wicket', {}).get('kind', '0')
#                     fielder = delivery[key].get('wicket', {}).get('fielders', '0')
#                     extra_run_type = list(delivery[key].get('extras', {}).keys())
#                     extra_run_type = extra_run_type and extra_run_type[0] or extra_run_type

#                     delivery_data = delivery_data.append({
#                         'match_id': match_id,
#                         'innings':inningKey,
#                         'info_Dates': info_dates,
#                         'teams': teams,
#                         'batting_team': batting_team,
#                         'ball': ball_of_match,
#                         'batsman': batsman,
#                         'non_striker':non_striker,
#                         'bowler': bowler,
#                         'total_runs': total_runs,
#                         'bat_runs': bat_runs,
#                         'extra_runs': extra_runs,
#                         'extra_run_type':extra_run_type,
#                         'player_dismissed': player_of_dismissed,
#                         'wicket_type':wicket_type,
#                         'fielder': fielder,
#                         'city': city,
#                         'venue': venue
#                     }, ignore_index=True)
                    

In [16]:
delivery_data.match_id.nunique()               

950

In [17]:
delivery_data

match_id      innings  info_Dates  \
0              1  1st innings  2017-04-05   
1              1  1st innings  2017-04-05   
2              1  1st innings  2017-04-05   
3              1  1st innings  2017-04-05   
4              1  1st innings  2017-04-05   
...          ...          ...         ...   
225949       950  2nd innings  2016-05-29   
225950       950  2nd innings  2016-05-29   
225951       950  2nd innings  2016-05-29   
225952       950  2nd innings  2016-05-29   
225953       950  2nd innings  2016-05-29   

                                                    teams  \
0       [Sunrisers Hyderabad, Royal Challengers Bangal...   
1       [Sunrisers Hyderabad, Royal Challengers Bangal...   
2       [Sunrisers Hyderabad, Royal Challengers Bangal...   
3       [Sunrisers Hyderabad, Royal Challengers Bangal...   
4       [Sunrisers Hyderabad, Royal Challengers Bangal...   
...                                                   ...   
225949  [Royal Challengers Bangalore, Sunrisers Hydera...   
225950  [Royal Challengers Bangalore, Sunrisers Hydera...   
225951  [Royal Challengers Bangalore, Sunrisers Hydera...   
225952  [Royal Challengers Bangalore, Sunrisers Hydera...   
225953  [Royal Challengers Bangalore, Sunrisers Hydera...   

                       batting_team  ball        batsman    non_striker  \
0               Sunrisers Hyderabad   0.1      DA Warner       S Dhawan   
1               Sunrisers Hyderabad   0.2      DA Warner       S Dhawan   
2               Sunrisers Hyderabad   0.3      DA Warner       S Dhawan   
3               Sunrisers Hyderabad   0.4      DA Warner       S Dhawan   
4               Sunrisers Hyderabad   0.5      DA Warner       S Dhawan   
...                             ...   ...            ...            ...   
225949  Royal Challengers Bangalore  19.2    Sachin Baby      CJ Jordan   
225950  Royal Challengers Bangalore  19.3    Sachin Baby      CJ Jordan   
225951  Royal Challengers Bangalore  19.4  Iqbal Abdulla    Sachin Baby   
225952  Royal Challengers Bangalore  19.5    Sachin Baby  Iqbal Abdulla   
225953  Royal Challengers Bangalore  19.6  Iqbal Abdulla    Sachin Baby   

          bowler  total_runs  bat_runs  extra_runs extra_run_type  \
0       TS Mills           0         0           0             []   
1       TS Mills           0         0           0             []   
2       TS Mills           4         4           0             []   
3       TS Mills           0         0           0             []   
4       TS Mills           2         0           2          wides   
...          ...         ...       ...         ...            ...   
225949   B Kumar           2         2           0             []   
225950   B Kumar           0         0           0             []   
225951   B Kumar           1         0           1        legbyes   
225952   B Kumar           1         1           0             []   
225953   B Kumar           4         4           0             []   

       player_dismissed wicket_type    fielder       city  \
0                     0           0          0  Hyderabad   
1                     0           0          0  Hyderabad   
2                     0           0          0  Hyderabad   
3                     0           0          0  Hyderabad   
4                     0           0          0  Hyderabad   
...                 ...         ...        ...        ...   
225949                0           0          0  Bangalore   
225950        CJ Jordan     run out  [NV Ojha]  Bangalore   
225951                0           0          0  Bangalore   
225952                0           0          0  Bangalore   
225953                0           0          0  Bangalore   

                                            venue  
0       Rajiv Gandhi International Stadium, Uppal  
1       Rajiv Gandhi International Stadium, Uppal  
2       Rajiv Gandhi International Stadium, Uppal  
3       Rajiv Gandhi International Stadium, Uppal  
4       Raj

In [21]:
pickle.dump(delivery_data,open('dataset_level3.pkl','wb'))